[REEF-1130] Worker side VortexAggregateFunction reception

afchung opened the pull request:
This addressed the issue by
  * Modified and added Avro classes to facilitate submission of Aggregate Functions to VortexWorkers.
  * Simple aggregation logic on VortexWorkers without AggregationPolicy.
  * Simple aggregation scenario and test.

JIRA:
  [REEF-1130](https://issues.apache.org/jira/browse/REEF-1130)

afchung commented:
@bgchun @yunseong @johnyangk Please have a look at this PR when you have time. Thanks!

afchung commented:
I'm not entirely sure how to solve the unchecked cast problem over serialization in this PR, any suggestions?

yunseong commented:
@afchung This is great! I'll have a look today. Thanks.

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexAggregateFuture.java:
```diff
@@ -64,11 +65,7 @@ public VortexAggregateFuture(final Executor executor,
    * @return the next aggregation result for the future, null if no more results.
    */
-  public synchronized AggregateResult get() throws InterruptedException {
-    if (taskletIdInputMap.isEmpty()) {
```
Could you elaborate why we don't need this check any more? Does't this method block when there is no more results, which is different from the documentation?

afchung replied:
I was previously using a `BlockingQueue` in a `synchronized` block with which I hold the same lock as I was putting into the queue. That was a bad Idea. This has been fixed by using a special lock for `get()`.

yunseong replied:
Thanks. Now it totally makes sense.

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexAggregateFuture.java:
```diff
@@ -78,12 +75,8 @@ public synchronized AggregateResult get() throws InterruptedException {
    */
-  public synchronized AggregateResult get(final long timeout,
+  public AggregateResult get(final long timeout,
                                           final TimeUnit timeUnit) throws InterruptedException, TimeoutException {
```
nit: indent

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexAggregateFuture.java:
```diff
@@ -93,6 +86,10 @@ public synchronized AggregateResult get(final long timeout,
   }
 
+  private synchronized boolean taskletIdInputMapIsEmpty() {
+    return taskletIdInputMap.isEmpty();
```
Adding `@GuardedBy` annotation to `taskletIdInputMap` will let us know that this object should be synchronized properly.

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexAggregateFuture.java:
```diff
@@ -182,6 +180,8 @@ public void run() {
       });
     }
 
+    LOG.log(Level.SEVERE, "PUT IN RESULT QUEUE!");
```
If this is not just for debugging, could you make the message informative, with more proper logging level?

afchung replied:
Sorry, it was for debugging.

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/TaskletAggregationRequest.java:
```diff
@@ -0,0 +1,100 @@
+   * @return Output of the function in a serialized form.
+   */
+  public byte[] executeAggregation(final List<TOutput> outputs) throws Exception {
+    LOGGER.log(Level.SEVERE, "Executing aggregation!");
```
This seems for debugging :-)

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/TaskletAggregationRequest.java:
```diff
@@ -0,0 +1,100 @@
+    final TOutput output = userAggregateFunction.call(outputs);
+    final Codec<TOutput> codec = userAggregateFunction.getOutputCodec();
+
+    LOGGER.log(Level.SEVERE, "Aggregate output is " + output);
```
This seems for debugging :-)

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexRequestor.java:
```diff
@@ -33,18 +33,30 @@
 
-  void send(final RunningTask reefTask, final VortexRequest vortexRequest) {
+  /**
+   * Sends a {@link VortexRequest} asyncrhonously to a {@link org.apache.reef.vortex.evaluator.VortexWorker}.
```
typo: `asyncrhonously` -> `asynchronously`

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexRequestor.java:
```diff
@@ -33,18 +33,30 @@
   }
+
+  /**
+   * Sends a {@link VortexRequest} syncrhonously to a {@link org.apache.reef.vortex.evaluator.VortexWorker}.
```
typo: `syncrhonously` -> `synchronously`

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/VortexWorkerManager.java:
```diff
@@ -41,17 +45,46 @@
+                                               final VortexFunction<TInput, TOutput> function) {
+    final TaskletAggregationRequest<TInput, TOutput> taskletAggregationRequest =
+        new TaskletAggregationRequest<>(aggregateFunctionId, aggregateFunction, function);
+    vortexRequestor.send(reefTask, taskletAggregationRequest);
```
`send()` is for guaranteeing that the message is sent before Tasklets are launched, right?
Adding a short comment why we want `send()` will be helpful to follow. :)

afchung replied:
Added comment.

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/evaluator/AggregateContainer.java:
```diff
@@ -0,0 +1,112 @@
+  private final Object stateLock = new Object();
+  private final TaskletAggregationRequest taskletAggregationRequest;
+  private final List<Pair<Integer, Object>> completedTasklets = new ArrayList<>();
+  private final List<Pair<Integer, Exception>> failedTasklets = new ArrayList<>();
```
What about adding `@GuardedBy("stateLock")`?

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/evaluator/VortexWorker.java:
```diff
@@ -53,15 +54,22 @@
 
   @Inject
   private VortexWorker(final HeartBeatTriggerManager heartBeatTriggerManager,
+                       final AggregateFunctionRepository aggregateFunctionRepository,
```
Isn't `AggregateFunctionRepository` for driver? Could you explain how this will be used in Workers?

afchung replied:
Added comment and moved `AggregateFunctionRepository` to `common`.

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/addition/Addition.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ * User's main function.
+ */
+final class Addition {
```
Honestly, it was a bit difficult for me to infer that this app counts the sum of 1s from the names: `Addition` and `Numbers`.

afchung replied:
Changed to `SumOnes` instead.

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/addition/Addition.java:
```diff
@@ -0,0 +1,64 @@
+        .build();
+
+    final VortexJobConf vortexJobConf = VortexJobConf.newBuilder()
+        .setJobName("Vortex_Example_AddOne")
```
The name should be consistent to the `Addition`

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/addition/AdditionAggregateFunction.java:
```diff
@@ -0,0 +1,48 @@
+  @Override
+  public Integer call(final List<Integer> taskletOutputs) throws VortexAggregateException {
+    int sum = 0;
+    for (int output : taskletOutputs) {
```
nit: final

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/addition/AdditionAggregateStart.java:
```diff
@@ -0,0 +1,82 @@
+import java.util.Vector;
+
+/**
+ * AddOne User Code Example.
```
This comment should also be consistent to `Addition`

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/addone/AddOne.java:
```diff
@@ -61,4 +61,4 @@ public static void main(final String[] args) {
   public static class Dimension implements Name<Integer> {
   }
-}
+}
```
Why don't we exclude this change, since it is not related to this PR?

yunseong commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/applications/vortex/addition/AdditionTestStart.java:
```diff
@@ -0,0 +1,85 @@
+import java.util.Vector;
+
+/**
+ * Test correctness of a addition aggregation test on Vortex.
```
nit: the article `a` is misused. Actually, I think we can rephrase this to something like `Test correctness of an aggregation function that adds integer outputs on Vortex`. How does it sound?

afchung replied:
Sounds good.

yunseong commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/applications/vortex/addition/AdditionTestStart.java:
```diff
@@ -0,0 +1,85 @@
+  }
+
+  /**
+   * Test correctness of a addition aggregation test on Vortex.
```
Same as above.

yunseong commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/applications/vortex/addone/AddOneTest.java:
```diff
@@ -85,7 +85,7 @@ public void testVortexAddOneCallback() {
         .set(VortexMasterConf.WORKER_CORES, 4)
         .set(VortexMasterConf.WORKER_CAPACITY, 2000)
-        .set(VortexMasterConf.VORTEX_START, AddOneTestStart.class)
+        .set(VortexMasterConf.VORTEX_START, AddOneCallbackTestStart.class)
```
Should we include it in this PR?

afchung replied:
@yunseong Opened 1149 for it, which is done in #789. Please have a look at that as well, thanks!

yunseong commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/applications/vortex/addone/AddOneTestStart.java:
```diff
@@ -64,4 +64,4 @@ public void start(final VortexThreadPool vortexThreadPool) {
     }
   }
-}
+}
```
Should we include it in this PR?

yunseong commented:
@afchung I've done a pass. This work is really great! Could you address some comments I left? Thanks!

afchung commented:
@yunseong Thanks for the careful review, looks like I haven't done a good job scrubbing my debug modifications before submitting the PR, sorry! I've addressed your comments, please have another look, thanks!

yunseong commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/addition/SumOnes.java:
```diff
@@ -0,0 +1,64 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.vortex.examples.addition;
```
I like `SumOnes` (actually I considered the same name :-))
One question: what do you think of the package name `addition`? I think it will be okay to leave it, but  `sumones` or `aggregation` sound better to me. Any thought?

yunseong commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/applications/vortex/addition/SumOnesTest.java:
```diff
@@ -0,0 +1,76 @@
+import org.junit.Test;
+
+/**
+ * Launch the Addition Vortex test.
```
This can also be changed as `SumOne` I guess.

yunseong commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/applications/vortex/addition/SumOnesTest.java:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  /**
+   * Run the Addition test.
```
`Addition` can be changed also

yunseong commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/applications/vortex/addition/SumOnesTest.java:
```diff
@@ -0,0 +1,76 @@
+   * Run the Addition test.
+   */
+  @Test
+  public void testVortexAddition() {
```
Addition -> SumOnes

yunseong commented on lang/java/reef-tests/src/test/java/org/apache/reef/tests/applications/vortex/addition/SumOnesTest.java:
```diff
@@ -0,0 +1,76 @@
+        .build();
+
+    final VortexJobConf vortexJobConf = VortexJobConf.newBuilder()
+        .setJobName("TEST_Vortex_AdditionTest")
```
Addition -> SumOnes

yunseong commented:
@afchung Thanks! I've left some minor comments. For your question, I don't have any concrete idea to solve the unchecked issues as well. 

@bgchun Could you kindly also take a look at this PR?

afchung commented:
@yunseong I've addressed your comments, please have another look. Thanks!

bgchun commented:
@afchung The test org.apache.reef.tests.applications.vortex.addition.SumOnesTest.testVortexSumOnes fails in Jenkins. Could you check it?

afchung commented:
@bgchun is there a way to get the log files on the machines? Would like to see if there is anything helpful there. Thanks!

afchung commented:
@bgchun Actually, I've found and fixed the bug, which was caused by a race condition in `VortexAggregateFuture.hasNext()`. I've addressed the problem by splitting up `AggregateResult` (which doesn't have `hasNext()`) and into `AggregateResultSynchronous` (which has `hasNext()`).

@bgchun and @yunseong please have another look, thanks!

bgchun commented:
@yunseong I quickly looked through the code. Could you take another look?

yunseong commented:
@bgchun @afchung Looks great. I'll merge this, and please let me have a minor change while merging (Exception -> VortexAggregateException in '@throws' javadoc in VortexAggregateFuture). Thanks!

afchung commented:
Thanks @yunseong!
